src/main/java/graphql/util/Traverser.java

```diff
@@ -0,0 +1,139 @@
+     * @return              Traverser instance
+     */
+    public static <T> Traverser<T> depthFirst (Function<? super T, ? extends List<T>> getChildren) {
+        return new Traverser<>(new TraverserStack<>(), getChildren);
+    }
+    
+    
+    /**
+     * Creates a standard Traverser suitable for breadth-first traversal
+     * 
+     * @param <T>           type of tree nodes to0 traverse
+     * @param getChildren   a function that obtains a list of children for a given tree node
+     * @return              Traverser instance
+     */
+    public static <T> Traverser<T> breadthFirst (Function<? super T, ? extends List<T>> getChildren) {
+        return new Traverser<>(new TraverserQueue<>(), getChildren);
+    }
+    
+    /**
+     * Resets the Traverser to the original state, so it can be re-used
+     */
+    public void reset () {
+        stack.clear();
+    }
+    
+    /**
+     * Starts traversal of a tree from a provided root using specified Visitor 
+     * and initial data to pass around Visitor's methods
+     * 
+     * @param <U>   type of data argument to Visitor's methods
+     * @param root  subtree root to start traversal from
+     * @param data  some data to pass across Visitor's methods. Visitor's methods
+     * can change that data to some other values of the same type or special Traverser
+     * markers {@link TraverserMarkers}
+     * @param visitor a Visitor object to be notified during traversal
+     * @return      some data produced by the last Visitor's method invoked
+     */
+    public <U> Object traverse (T root, U data, TraverserVisitor<? super T, ? super U> visitor) {
+        return traverse(Collections.singleton(root), data, visitor);
+    }
+    
+    
+    /**
+     * Starts traversal of a tree from a provided roots using specified Visitor 
+     * and initial data to pass around Visitor's methods
+     * 
+     * @param <U>   type of data argument to Visitor's methods
+     * @param roots  multiple subtree roots to start traversal from
+     * @param data  some data to pass across Visitor's methods. Visitor's methods
+     * can change that data to some other values of the same type or special Traverser
+     * markers {@link TraverserMarkers}
+     * @param visitor a Visitor object to be notified during traversal
+     * @return      some data produced by the last Visitor's method invoked
+     */
+    public <U> Object traverse (Collection<T> roots, U data, TraverserVisitor<? super T, ? super U> visitor) {
+        Objects.requireNonNull(roots);
+        Objects.requireNonNull(visitor);
+        
+        stack.addAll(roots);
+        
+        Object d = data;
+        while (!(stack.isEmpty() || (d = traverseOne((TraverserVisitor<T, U>)visitor, (U)d)) == TraverserMarkers.QUIT));
+        
+        return d;
+    }
+    
+    protected <U> Object traverseOne (TraverserVisitor<T, U> visitor, U data) {
```

Comment:
why is this protected and not private?

Reply:
For possible subclassing.
For instance, it is very easy to create PreOrder, PostOrder or BreadthFirst iterators using this mechanism. One only need to call `traverseOne` as iterator's advance method and use external loop while condition as iterator's method to check if there are more elements to traverse.
I did not want to put this in this class for the reasons not to abuse its purpose. When iterators are required, one cold either put them right in  this class or subclass it just for the purpose of adding that behavior.